File: BackTracking/Combinations.cpp
//Problem link:- https://leetcode.com/problems/combinations/

/* Given two integers n and k, return all possible combinations of k numbers out of 1 ... n. */

#include<bits/stdc++.h>
using namespace std;

class Solution {
public:
    void helper(vector<vector<int>>& result,vector<int> & path,int num,int n, int k){
        if (path.size() == k){
            result.push_back(path);
            return;
        }
        if (num > n)
            return;
        path.push_back(num);
        helper(result,path,num+1,n,k);
        path.pop_back();
        helper(result,path,num+1,n,k);
    }
    
    vector<vector<int>> combine(int n, int k) {
        vector<vector<int>> result;
        vector<int> path;
        helper(result,path,1,n,k);
        return result;
    }
};

File: BackTracking/Combinations.py
# Problem link:- https://leetcode.com/problems/combinations/

class Solution:
    def combine(self, n: int, k: int) -> List[List[int]]:
        result = []
        if n == 1 and k == 1:
            return [[1]]
        arr = [x for x in range(1,n+1)]
        self.helper(n,k,arr,0,0,result,[])
        return result
    
    def helper(self,n,k,arr,index,cnt,result,path):
        if cnt == k:
            result.append(path)
            return 
        if index >= len(arr):
            return 
        self.helper(n,k,arr,index+1,cnt,result,path)
        self.helper(n,k,arr,index+1,cnt+1,result,path+[arr[index]])